### DISCUSSION TITLE
container.play() plays only once

### DISCUSSION THREAD

**Discussion by jkazimierczak** [matteobruni/tsparticles#5009] [Q&A] [answered]
I'm using the tsparticles with React component. When I try to invoke `play()` on container ref, the first time the effect plays fine, but the second time it doesn't play.
```json
"dependencies": {
  "react": "^18.2.0",
  "react-dom": "^18.2.0",
  "react-tsparticles": "^2.10.1",
  "tsparticles": "^2.10.1",
  "tsparticles-engine": "^2.10.1"
},
```
Here's a codesanbox with the code:
https://codesandbox.io/p/sandbox/quizzical-microservice-wt3psp

**Comment by matteobruni** [comment]
Based on your config is working fine, maybe you wanted to add more other emitters. The one set in the config will despawn and the container is not stopped because of that, and it’s intended this way.

**Reply by jkazimierczak** [reply]
Unless I misunderstand how this works, I don't need more emitters to play the desired effect (confetti). However I need to recreate the emitter after the one specified in config is destroyed, right?

If I found it correctly, a new emitter can be constructed with `EmitterInstance`, but how do I add it to the existing container then? The constructor also has an `engine` parameter, which is only available inside of the `useCallback`.
https://particles.js.org/docs/classes/tsParticles_Emitters_Plugin.EmitterInstance.EmitterInstance.html

**Reply by matteobruni** [reply]
https://github.com/matteobruni/tsparticles/discussions/4931#discussioncomment-5494349

**Reply by jkazimierczak** [reply, marked as answer]
Thank you! I managed to solve it with that hint, although it required a bit of searching. 

The only thing I'm missing is how to solve typing issues, as if I replace `Container` with `EmitterContainer` then I get an error for `loaded` property of Particles component:
```
TS2322: Type '(container: Container | undefined) => Promise<void>' is not assignable to type '(container?: Container | undefined) => Promise<void>'.   Types of parameters 'container' and 'container' are incompatible.
```
If I keep the `Container` then I get an error inside of particlesLoaded callback:
```ts
const particlesLoaded = useCallback(async (container: Container | undefined) => {
  // TS2339: Property 'removeEmitter' does not exist on type 'Container'.   Property 'removeEmitter' does not exist on type 'Element'
  container?.removeEmitter("customConfetti");
}, []);
```

So for the future:
- working demo: https://codesandbox.io/p/sandbox/particlests-play-issue-forked-52kcfl?file=%2Fpackage.json%3A29%2C1
- step-by-step below

## Config
Extract the `emitters` object to a separate variable:
```ts
export const confettiEmitter: RecursivePartial<IEmitter> = {
  autoPlay: true,
  fill: true,
  life: {
    wait: false,
    count: 8,
    delay: 0.2,
    duration: 0.1,
  },
  rate: {
    quantity: 150,
    delay: 0.1,
  },
  shape: "square",
  startCount: 0,
  size: {
    mode: "percent",
    height: 0,
    width: 0,
  },
  name: "customConfetti",
  particles: {},
};
```
and add a reference to it in main options object:
```ts
export const confettiConfig: ISourceOptions = {
  // other config
  emitters: confettiEmitter,
};
```

## Usage
1. Replace `Container` with `EmitterContainer` for typing (necessary to add `tsparticles-plugin-emitters` to dependencies).
2. Add `containerRef.current?.addEmitter(confettiEmitter);` before using `containerRef.current?.play();`.

**Reply by jkazimierczak** [reply]
After further investigation with a debugger I see that the container in the useCallback is in fact of type Container. Despite what `Container` type says, at runtime it has all the methods as specified in `EmitterContainer` type.

I'm not an expert in TS, but there's clearly a type issue, either in useCallback or in the component props itself. I would suspect that the type specified in an interface IParticlesProps.d.ts is too narrow:
```ts
export interface IParticlesProps {
    id?: string;
    width?: string;
    height?: string;
    options?: ISourceOptions;
    params?: ISourceOptions;
    url?: string;
    style?: CSSProperties;
    className?: string;
    canvasClassName?: string;
    container?: RefObject<Container>;
    init?: (engine: Engine) => Promise<void>;
    loaded?: (container?: Container) => Promise<void>;
}
```

**Reply by matteobruni** [reply]
The type is correct, it needs to be casted since the EmitterContainer is inside a plugin that could be not imported by some users. Thanks for the additional steps 👍

**Reply by jkazimierczak** [reply]
Thanks you also for help! Much appreciated!

**Reply by jkazimierczak** [reply]
Just one more thing for anyone wondering in the future - the syntax is as follows:
```ts
const particlesLoaded = useCallback(async (container: Container | undefined) => {
  (container as EmitterContainer)?.removeEmitter("customConfetti");
}, []);

function showConfetti() {
  (containerRef.current as EmitterContainer)?.addEmitter(confettiEmitter);
  containerRef.current?.play();
}
```

**Reply by matteobruni** [reply]
You can use emitters without names, that will prevent the `removeEmitter` call

**Reply by jkazimierczak** [reply]
I tested that out, but this results in the effect being initially played, which I don't want.

However, `removeEmitter` can be removed if the `emitters` in the options object are set to an empty array. This way it turns out that I don't need `particlesLoaded` useCallback, and the prop can be removed altogether.
